2025/day2.md

```diff
@@ -1,46 +1,200 @@
-# Day 2 – [Lab Title Here]
+# Day 2 – Linux Kernel Modules: Character Drivers, GPIO, and User-Space Communication
 
 Slides: [day2_slides](./day2_slides)
 
 ## Description
 
-_Provide a short description of what this lab session is about. Mention the scope and practical goals
-for the day, what students are expected to understand or achieve by the end of this lab._
+This lab introduces key concepts in Linux kernel module development, with a focus on character drivers and the interaction between user space and kernel space. Through a sequence of practical exercises, the lab explores the use of file operations, ioctl, and GPIO-based hardware control. The final goal is to build a solid understanding of how user applications can safely communicate with kernel modules to perform data exchange and hardware manipulation.
 
 ---
 
 ## Topics Covered
 
-- _List bullet points summarizing key technical concepts introduced during this session._
-- _These can include kernel mechanisms, APIs, interfaces, or subsystems discussed._
-- _Use this as a quick reference or learning summary._
+- Writing Linux kernel modules
+- Implementing character device drivers (`open`, `read`, `write`, `release`)
+- Data exchange between user space and kernel space
+- Device registration: static and dynamic methods
+- Using `ioctl` for custom commands
+- GPIO handling in the Linux kernel
+- Controlling LEDs (on/off, blinking) through device drivers
 
 ---
 
 ## Lab Exercises
 
-### Exercise 1: [Exercise Title]
-
-_Goal: Clearly describe the objective of this exercise and what the student is expected to build, test,
-or analyze._
-
+Before starting this lab, please make sure you enable **CONFIG_LKSS_DRIVERS_LAB2**.
+
+**HINT**: Use `make menuconfig`
+
+### Exercise 1: Simple Character Driver – File Operations
+
+**Objective**: Implement a basic character device driver (simple_char.c) that supports `open`, `read`, `write`, and `release` operations.
+The module is inserted using `modprobe`, and the device node is manually created using `mknod`.
+Test communication between a user-space application and the device, verifying basic file operations and kernel logs using `dmesg`.
+
+**Steps**:
+1. Create the Kernel Module
+- Use the provided `simple_char.c` template. Look into `drivers/lkss/lab2/ex1`
+- Implement the four file operations: `open()`, `read()`, `write()`, `release()`.
+2. Register the Device
+- In `char_init()` function, use `register_chrdev()` to register a major number and associate it with the file operations, 
```

Comment:
In the `char_init()` function.

use full stop instead of comma at the end of the sentence